openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/changeEvent/AbstractEventConsumer.java

```diff
@@ -0,0 +1,316 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.openmetadata.service.apps.bundles.changeEvent;
+
+import static org.openmetadata.schema.entity.events.SubscriptionStatus.Status.ACTIVE;
+import static org.openmetadata.schema.entity.events.SubscriptionStatus.Status.AWAITING_RETRY;
+import static org.openmetadata.schema.entity.events.SubscriptionStatus.Status.FAILED;
+import static org.openmetadata.service.events.subscription.AlertUtil.getFilteredEvent;
+import static org.openmetadata.service.events.subscription.AlertUtil.getInitialAlertOffsetFromDb;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.openmetadata.schema.entity.events.AlertMetrics;
+import org.openmetadata.schema.entity.events.EventSubscription;
+import org.openmetadata.schema.entity.events.EventSubscriptionOffset;
+import org.openmetadata.schema.entity.events.FailedEvent;
+import org.openmetadata.schema.entity.events.SubscriptionStatus;
+import org.openmetadata.schema.type.ChangeEvent;
+import org.openmetadata.service.Entity;
+import org.openmetadata.service.events.errors.EventPublisherException;
+import org.openmetadata.service.events.subscription.AlertUtil;
+import org.openmetadata.service.util.JsonUtils;
+import org.quartz.DisallowConcurrentExecution;
+import org.quartz.Job;
+import org.quartz.JobDetail;
+import org.quartz.JobExecutionContext;
+import org.quartz.JobExecutionException;
+import org.quartz.PersistJobDataAfterExecution;
+
+@Slf4j
+@DisallowConcurrentExecution
+@PersistJobDataAfterExecution
+public abstract class AbstractEventConsumer implements Consumer<ChangeEvent>, Job {
+  public static final String ALERT_OFFSET_KEY = "alertOffsetKey";
+  public static final String ALERT_INFO_KEY = "alertInfoKey";
+  public static final String OFFSET_EXTENSION = "eventSubscription.Offset";
+  public static final String METRICS_EXTENSION = "eventSubscription.metrics";
+  public static final String FAILED_EVENT_EXTENSION = "eventSubscription.failedEvent";
+  private int offset = -1;
+  private AlertMetrics alertMetrics;
+
+  @Getter @Setter private JobDetail jobDetail;
+  protected EventSubscription eventSubscription;
+
+  protected AbstractEventConsumer() {}
+
+  private void init(JobExecutionContext context) {
+    EventSubscription sub =
+        (EventSubscription) context.getJobDetail().getJobDataMap().get(ALERT_INFO_KEY);
+    this.jobDetail = context.getJobDetail();
+    this.eventSubscription = sub;
+    this.offset = loadInitialOffset(context);
+    this.alertMetrics = loadInitialMetrics();
+    this.doInit(context);
+  }
+
+  protected abstract void doInit(JobExecutionContext context);
+
+  protected void sendAlert(ChangeEvent event) throws EventPublisherException {
```

Comment:
can we move this out of it and create another interface that way any consumer who wants to implement sendAlert they extends this class and implement the interface for sending alert

Reply:
done